### ISSUE TITLE
New connections with same priority not accepted in place of old ones until long timeout expires (>5 minutes)

### ISSUE THREAD

**Issue by olifre** [syncthing/syncthing#9337]
### Does your log mention database corruption?

No. 

### Include required information
- Syncthing v1.27.2
- Tested between Gentoo Linux and Debian, and Android and Debian

### Test setup
- One client (A) with a VPN connection to a given network
- The other client (B) lives locally in the given network

To reproduce:
1. Client (A) connects to the VPN, and subsequently, after discovery / redialing, upgrades the Syncthing connection to a "LAN" connection to Client (B), as expected. 
2. Client (A) disconnects from the VPN, and then enters the actual LAN, e.g. by connecting to it directly via WiFi. Alternatively, it uses a different VPN / IP address within the LAN. 
3. Client (B) remains "connected" to Client (A), via the old address, and rejects any new connection with `rejected: already connected to this device`. Since the initial connection was a LAN connection, priorities are the same. 

Effect:
Reconnection of Client (A) is blocked by Client (B) until a long timeout (>5 minutes) has expired. Maybe that's a TCP timeout, actually. 

## Context
I think this was initially fixed in #5224, but seems to have regressed afterwards, the logic added there seems gone in the current code. Of course, an alternative approach with current Syncthing might be to accept an additional connection of the same priority from the new address (since multiple connections are supported) and wait until the old one expires after the long timeout. 
Note my tests were done with default configuration, i.e. "1 connection" configured for all clients.

**Comment by calmh** [comment]
Yeah, that's [protocol.ReceiveTimeout](https://github.com/syncthing/syncthing/blob/8edd67a569594bec06c395041acb7eb18f8e3b13/lib/protocol/protocol.go#L239) in action. The root cause here is that the connection is never closed, the other side just disappears from the network, so the timeout is the only thing that reaps the connection in the end.

I'm unsure if "replacing" the connection is a good thing nowadays in the multiple-connection world. It was an easier thing to reason about before, when there was only ever supposed to be one.

**Comment by olifre** [comment]
> Yeah, that's [protocol.ReceiveTimeout](https://github.com/syncthing/syncthing/blob/8edd67a569594bec06c395041acb7eb18f8e3b13/lib/protocol/protocol.go#L239) in action.

That makes a lot of sense, I was looking for some timeout close to 5 minutes, but apparently missed this one. Indeed, in such "roaming" cases, connections are often never closed. 

> I'm unsure if "replacing" the connection is a good thing nowadays in the multiple-connection world

I fully agree. I don't have the magic bullet idea, but maybe something like accepting an "extra connection" in case it is of the same priority, but with source different address than existing ones (i.e. assumption of roaming), and then just letting `ReceiveTimeout` do its thing for the remaining connections would be reasonable? 
In the worst case, this could lead to cases with one more active connection than configured, but it would help in roaming cases. But maybe such a logic has to wait until >1 connections becomes a default (in case that is planned)?

**Comment by imsodin** [comment]
> There is a simple negotiation scheme to let two sides agree on the number of connections in use, but I don't think it's foolproof and it would be less than optimal if two sides with slightly different ideas on the number of connections to use would cause continuous connection replacement churn.

What about only replacing the connection if the remote announced that they only want a single connection? I don't see any risk of continuous connection replacement churn in that scenario.  
It should also be safe if the local instance wants only a single connection, as the remote should be aware, but I guess not relying on that is what you referred to with "don't think it's foolproof".

**Comment by calmh** [comment]
I guess 🤷

**Comment by lheckemann** [comment]
Would reducing `protocol.ReceiveTimeout` cause any problems?

My use case: My phone is often on my 2.4GHz WiFi network (which reaches into all the corners of my apartment). I sometimes switch it to the 5GHz network (which is much faster) for the express purpose of syncing more quickly. Having to wait 5 minutes before the server accepts the connection negates that benefit though, and manually going into the UI on the other end to pause/unpause the phone device is pretty inconvenient. I guess reducing ReceiveTimeout would help here, or is there a better way to address the problem in my scenario?
